[Doc] Installation guide & Integration guide (#142)
3scale/zync: INTEGRATE.md

# Local integration between Porta/System with Keycloak through Zync
Install these 3 components to have a complete integration of managed SSO applications in Porta.

## Install Porta
Follow [the installation guide](https://github.com/3scale/porta/blob/master/INSTALL.md).

## Install Zync
Follow [the installation guide](https://github.com/3scale/zync/blob/master/INSTALL.md).

## Install Keycloak
To have a keycloak running locally, do `docker run --name keycloak -d -p 8080:8080 -e KEYCLOAK_USER=admin -e KEYCLOAK_PASSWORD=p jboss/keycloak`

> Finally [configure Red Hat Single Sign-On](https://access.redhat.com/documentation/en-us/red_hat_3scale_api_management/2.4/html/api_authentication/openid-connect).

Reviewer: I think this is redundant with the last section

Author: @hallelujah It is not the same, this is `2.3.1. Configure Red Hat Single Sign-On` and the other one is `2.3.2. Configure 3scale` and it needs to be done in this order 🤔

## Run Zync

> From the folder where Zync is installed, run: `ZYNC_AUTHENTICATION_TOKEN=zynctoken bundle exec rails server -p 5000` to get zync running in the port 5000.

Reviewer: The token can be replaced right?

## Run Porta
These configurations are needed:
  - In `config/rolling_updates.yml`, make sure to have `apicast_v2: true` and `apicast_oidc: true`.
  - In `config/sandbox_proxy.yml`, make sure to have `apicast_registry_url: <%= ENV.fetch('APICAST_REGISTRY_URL', '<Insert a URL that contains policies in JSON. It can be am empty JSON>') %>`

Run porta with `UNICORN_WORKERS=8 ZYNC_ENDPOINT=http://localhost:5000 ZYNC_AUTHENTICATION_TOKEN=zynctoken bundle exec rails server -b 0.0.0.0`

### Run Porta's sidekiq

> It works in Background in the `low` queue so do this: `ZYNC_ENDPOINT=http://localhost:5000 ZYNC_AUTHENTICATION_TOKEN=zynctoken bundle exec sidekiq -q default -q mailers -q low -q critical`

Reviewer: Sidekiq Zync Worker process jobs from the `low` queue.

Author: @hallelujah What about this then?
```suggestion
Sidekiq Zync Worker process jobs from the `low` queue so do this: `ZYNC_ENDPOINT=http://localhost:5000 ZYNC_AUTHENTICATION_TOKEN=zynctoken bundle exec sidekiq -q low`
```

## Configure 3scale to use Keycloak

> Follow the documentation to [configure 3scale](https://access.redhat.com/documentation/en-us/red_hat_3scale_api_management/2.4/html/api_authentication/openid-connect).

Reviewer 2: _https://access.redhat.com/documentation/en-us/{DocInfoProductName}/{ProductVersion}/html/api_authentication/openid-connect_

Between docs you need full URLs with placeholders for product name ({DocInfoProductName}) and doc version ({ProductVersion}) used as variables, as in the following example: https://access.redhat.com/documentation/en-us/{DocInfoProductName}/{ProductVersion}/html-single/deployment_options/#apicast_policies[Policies]

Author: As spoken, it will be done as another task because it is for an internal audience and that way we are not blocked.
[THREESCALE-2075 - Autoupdate links of internal documentation](https://issues.jboss.org/browse/THREESCALE-2075)

> # Local integration between Porta/System with Keycloak through Zync

Reviewer 3: I think maybe an opening sentence about what this integration solves would make sense here. 

e.g. with all 3 components installed, you can do X, Y and Z... and maybe the role each plays in that. 

(not long text though - just a tl;dr kind of thing)

## Install Porta
Follow [the installation guide](https://github.com/3scale/porta/blob/master/INSTALL.md).

## Install Zync
Follow [the installation guide](https://github.com/3scale/zync/blob/master/INSTALL.md).

## Install Keycloak
You can have a keycloak running locally with `docker run --name keycloak -d -p 8080:8080 -e KEYCLOAK_USER=admin -e KEYCLOAK_PASSWORD=p jboss/keycloak`
So that docker container with keycloak will be accessible from your browser in `http://localhost:8080`.

After that, you will need to follow the instructions to [configure Red Hat Single Sign-On](https://access.redhat.com/documentation/en-us/red_hat_3scale_api_management/2.4/html/api_authentication/openid-connect).

## Run Zync
From the folder where you have Zync installed, run: `ZYNC_AUTHENTICATION_TOKEN=zynctoken bundle exec rails server -p 5000` so you will get zync running in the port 5000.

> ## Run Porta

Reviewer 3: i meant just a numbered list of steps under this heading yes, but not a major thing... nvm.

> If your porta is already running, stop it, and if you have `spring`, also do `spring stop`.

Reviewer 3: does that make summer come early ?

Author: Then no need to do it, this is only for those who have Spring running. It is in the instructions to install Porta/System as optional:
https://github.com/3scale/porta/blob/master/INSTALL.md#spring-optional

> Spring is a Rails application preloader. It speeds up development by keeping your application running in the background so you don't need to boot it every time you run a test, rake task or migration.
> This is not required but still recommended. Install it via gem: `gem install spring -v 2.0.0`

Reviewer: I think this is troubleshooting not really installation. And only Porta related. Is there any need to mention details of Porta in Zync?

You will need these configurations:
  - In `config/rolling_updates.yml`, make sure you have `apicast_v2: true` and `apicast_oidc: true`.

>   - In `config/sandbox_proxy.yml`, make sure you have `apicast_registry_url: <%= ENV.fetch('APICAST_REGISTRY_URL', '<Insert a URL that contains policies in JSON. It can be am empty JSON>') %>`

Reviewer: OK this is porta issue.
I think by default we should not need this for running porta in development mode. Now that we have custom policies. Opening an issue for that?

Author: Better, yes, I will open an issue for that in Jira, but this is definitely out of scope for merging this 😛

Author: @hallelujah Created in [THREESCALE-2060](https://issues.jboss.org/browse/THREESCALE-2060) 😄

# Local integration between Porta/System with Keycloak through Zync

> With these 3 components installed, you will be able to synchronize your applications in Porta with your Keycloak (Red Hat SSO)

Reviewer: I would avoid using `you` in instructions but be more imperative and using the present tense

Proposition:

```suggestion
Install these 3 components to have a complete integration of managed SSO application in Porta
```

## Install Porta
Follow [the installation guide](https://github.com/3scale/porta/blob/master/INSTALL.md).

## Install Zync
Follow [the installation guide](https://github.com/3scale/zync/blob/master/INSTALL.md).

## Install Keycloak
You can have a keycloak running locally with `docker run --name keycloak -d -p 8080:8080 -e KEYCLOAK_USER=admin -e KEYCLOAK_PASSWORD=p jboss/keycloak`

> So that docker container with keycloak will be accessible from your browser in `http://localhost:8080`.

Reviewer: I do not know if there is a need to write this sentence. Docker user knows what that means. But you can say,

Note: Do not forget to map ports of your docker container with localhost `-p 8080:8080`

Author: @hallelujah Then I will simply remove the sentence, because I already said in the previous line the mapping of the ports... I don't need to say it twice 😄

> After that, you will need to follow the instructions to [configure Red Hat Single Sign-On](https://access.redhat.com/documentation/en-us/red_hat_3scale_api_management/2.4/html/api_authentication/openid-connect).

Reviewer: > Finally configure Red Hat Single Sign on

## Run Zync
From the folder where you have Zync installed, run: `ZYNC_AUTHENTICATION_TOKEN=zynctoken bundle exec rails server -p 5000` so you will get zync running in the port 5000.

## Run Porta
If your porta is already running, stop it, and if you have `spring` running, also do `spring stop`.
You will need these configurations:

>   - In `config/rolling_updates.yml`, make sure you have `apicast_v2: true` and `apicast_oidc: true`.

Reviewer: I think we should make those rolling updates default in Porta now so you can remove those instructions

Author: @hallelujah Do you mean in the Porta repo making a PR with these configs in `config/examples/rolling_updates.yml` ?

Reviewer: No I mean removing the protection of rolling updates on those feature.
They should be default now

Author: @hallelujah I've just created this https://issues.jboss.org/browse/THREESCALE-2059
So this is not a blocker to merge this for now and I can do the more urgent tasks first and people can have some instructions to use Zync.

# Local integration between Porta/System with Keycloak through Zync
Install these 3 components to have a complete integration of managed SSO applications in Porta.

1. Install Porta
Follow [the installation guide](https://github.com/3scale/porta/blob/master/INSTALL.md).

2. Install Zync
Follow [the installation guide](https://github.com/3scale/zync/blob/master/INSTALL.md).

> 

Reviewer 2: Instruction in one line.
`commands` in the next line(s).

Author: outdated, old version 😛
